### ISSUE TITLE
Remote data loading support

### ISSUE THREAD

**Issue by andrewfan** [miguelcobain/ember-selectize#18]
What do you think about this feature?

**Comment by miguelcobain** [comment]
I'm not sure how to implement this feature in an Ember context.
Why is that? Because, in an Ember app, you generally delegate "data related" things to another layer.

I remind you that you can easily implement this yourself. Ember-selectize updates a `filter` property with the text you type in the input box. This allows you to define a computed property based on this filter. Let me illustrate what I mean. Imagine you're in a template, and you want the user to select a street from a huge list loaded remotely. In your template, you would have:

``` handlebars
{{view Ember.Selectize contentBinding="streets"
    optionValuePath="content.id"
    optionLabelPath="content.label"
    selectionBinding="place"
    filterBinding="streetFilter"}}
```

Notice the `filterBinding="streetFilter"`. Now you can define a computed property depending on this filter that returns an array. In your controller you would have:

``` javascript
  streets:function(){
    var streetFilter = this.get('streetFilter'); //the filter string you entered
    var modelPlace = this.get('content.place'); //the model's current street, may be empty

    //Now you can load records (based on the filter)
    //with Ember-Data or whatever you use for your models

    //return an array.
    //Important. Remember that the model's street itself must be contained in this array.
    //Otherwise, ember-selectize can't render your option,
    //as it isn't on the possible options available (content array)
  }.property('streetFilter','place')
```

You can even throttle this computed property, and I think you get a decent flexibility.

The advantage of this approach is that you have a more clear separation of concerns.
The disadvantage I see is that you don't see Selectize's spinner, indicating that a remote loading is in progress. You basically swap selectize's options. Selectize doesn't "know" how or why they have been swapped. But maybe this issue can be resolved manually.

I'll leave this issue open for discussion. Maybe someone has a clear idea of how to better implement this. Hope you found this useful.

**Comment by andrewfan** [comment]
selectize.js supports remote data loading out from the box. if you set `load` method in selectize options, selectize will do remote calls. may be we can use it?

**Comment by miguelcobain** [comment]
The question is if you want selectize to do that.
If you want to show your own models in selectize, probably not because you will want to have your authentication, and custom logic.
But maybe for other sources like github, or some not so complicated things, it may be useful.

Any thoughts on how to provide your own load function?
By extending Ember.Selectize class and defining your own load function in it?

**Comment by andrewfan** [comment]
Yes and no. We can do extending, but we should pass `load` method in selectize options (if this option is specified selectize.js will toggle `loading` class each time user types). 
Also i found 1 more async usecase where ember-selectize works weird. imagine we have:

```
Person = Ember.Object.extend({
    contact: null  //relation to other model
});
```

I want to provide viewing\editing functionality and i have `person` with defined `contact`. When pages loads select is empty, because as you mentioned above if `selection` is defined and `content` is empty or does not contain `selection` select wont work. But i cant load all contacts from server each time i want to view\update single contact.
Should we add `selection` to `content` manually at https://github.com/brianreavis/selectize.js/blob/master/src/selectize.js#L298, or should i do it somewhere outside view or should i forget about using ember-selectize in async situations?

**Comment by miguelcobain** [comment]
> We can do extending, but we should pass load method in selectize options (if this option is specified selectize.js will toggle loading class each time user types).

That's what I meant. We could, by default, pass a `load` function if present in Ember.Selectize class (or subclass) into selectize's options. If such a function isn't present, we wouldn't pass that parameter.

Regarding that scenario, what I do is to make sure that my `content` **always** contains at least the `selection` (the selected object). This way you don't need to load all your contacts contacts. You just need to indicate that the selection is a possible option by including it in your `content` array.

Basically, `content` is a collection of possible options. Anything **not** in it is an invalid option. This is necessary in order to distinguish valid from invalid situations and it reflects the way selectize works.

Hope I helped.

**Comment by andrewfan** [comment]
> Regarding that scenario, what I do is to make sure that my content always contains at least the
>  selection (the selected object). This way you don't need to load all your contacts contacts. You just 
> need to indicate that the selection is a possible option by including it in your content array.

So why not to do it automatically? We have a perfect place for it
https://github.com/miguelcobain/ember-selectize/blob/master/src/ember.selectize.js#L179. Maybe it should depend on `load` param, to do it only when we work in `async` mode:

```
_selectionDidChange : Ember.observer(function() {
  if(!this.inDOM) return;

  //also we can add if (this.load && ...
  if(!this.get('content').findBy(get(this,'_valuePath'),get(this.get('selection'),get(this,'_valuePath'))))
  {
  this.get('content').addObject(this.get('selection'));
 }
 ...
```

What do you think about it? If you want a can do a PR with `load` param implementation and method example

**Comment by miguelcobain** [comment]
One of the goals of this project is to be a drop-in replacement for `Ember.Select`. If I remember correctly, this "limitation" (debatable) is in Ember.Select itself. So, in order to pass the same tests, we need this. (Tests are incomplete)

Imagine that an invalid value (say, `null` or worse), ends up in `selection`. Do we want ember-selectize to _automatically_ recognize it as a valid option? At a first sight this seems a bit fragile to me.

If you want to make your own experiments, remember that you can always subclass `Ember.Selectize` and override any method.

Your link points to `_selectionDidChange`, but the code you wrote has `_selectionWillChange`. Didn't understand exactly where you suggested to code.

However, if I were to implement it like you suggested, I would probably use a computed property that would be basically the concatenation of `content` with `[selection]`. Then, the same logic would apply to this array.

**Comment by andrewfan** [comment]
Sorry for incorrect link. Thanks for help.

**Comment by miguelcobain** [comment]
You're welcome. Hope I helped. Feel free to question anything.

I'll leave this open to get other people's opinions on the remote loading feature.

**Comment by miguelcobain** [comment]
This issue led me to implement a basic loading feature for ember selectize.
Please check commit f58ae59d92e2d9e7f1784a72333a1dc8389694ba.

It basically adds a loading css class, just like selectize does it.

You can use it like:

``` handlebars
{{ view Ember.Selectize id="type"
  contentBinding="controller.types"
  optionValuePath="content.id"
  optionLabelPath="content.name"
  selectionBinding="model.type"
  placeholder="Select an option"
  loadingBinding="controller.types.isPending"}}
```

Then customize your css to whatever the appearance you like, just like you need to do with selectize.
If you're using font-awesome, here are some styles you can use:

``` css
.selectize-control::before {
  font-family: 'FontAwesome';
  -moz-transition: opacity 0.2s;
  -webkit-transition: opacity 0.2s;
  transition: opacity 0.2s;
  content: ' ';
  z-index: 2;
  position: absolute;
  display: block;
  top: 50%;
  right: 34px;
  width: 16px;
  height: 16px;
  margin: -8px 0 0 0;
  line-height: 16px;
  text-align: center;
  content: "\f110";
  opacity: 0;

  -webkit-animation: spin 2s infinite linear;
  -moz-animation: spin 2s infinite linear;
  -o-animation: spin 2s infinite linear;
  animation: spin 2s infinite linear;
}

.selectize-control.loading::before {
  opacity: 0.4;
}
```

(inspired by selectize demo, but using font-awesome)

Also, this goes along very well with Ember Data's Promise Objects and Promise Arrays. Just bind to the `isPending` property of them. However, any boolean will do!

The leads me to the conclusion that we should not address selectize remote loading feature here. That is for an Ember app to do, not a combo box component. The context is different here.

The demo page should be updated with this example.

Let me know of any issues you might have.
